# Mod 2 Review

## Important note:

Questions 3, 4a and 4b are on content about `Multiple Regression`

Successful students will likely read ahead in the Canvas material and attempt to answer them *before* Wednesday (which is when we will receive lecture on the `Multiple Regression` material)

In [2]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from statsmodels.formula.api import ols

pd.set_option('display.max_columns', 100)

### Check Your Data … Quickly
The first thing you want to do when you get a new dataset, is to quickly to verify the contents with the .head() method.

In [5]:
df = pd.read_csv('movie_metadata.csv')
print(df.shape)
df.head()
df.info()
df.describe()

(5043, 28)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
color                        5024 non-null object
director_name                4939 non-null object
num_critic_for_reviews       4993 non-null float64
duration                     5028 non-null float64
director_facebook_likes      4939 non-null float64
actor_3_facebook_likes       5020 non-null float64
actor_2_name                 5030 non-null object
actor_1_facebook_likes       5036 non-null float64
gross                        4159 non-null float64
genres                       5043 non-null object
actor_1_name                 5036 non-null object
movie_title                  5043 non-null object
num_voted_users              5043 non-null int64
cast_total_facebook_likes    5043 non-null int64
actor_3_name                 5020 non-null object
facenumber_in_poster         5030 non-null float64
plot_keywords                4890 non-null object
movie_imdb_link            

num_critic_for_reviews     duration  director_facebook_likes  \
count             4993.000000  5028.000000              4939.000000   
mean               140.194272   107.201074               686.509212   
std                121.601675    25.197441              2813.328607   
min                  1.000000     7.000000                 0.000000   
25%                 50.000000    93.000000                 7.000000   
50%                110.000000   103.000000                49.000000   
75%                195.000000   118.000000               194.500000   
max                813.000000   511.000000             23000.000000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross  \
count             5020.000000             5036.000000  4.159000e+03   
mean               645.009761             6560.047061  4.846841e+07   
std               1665.041728            15020.759120  6.845299e+07   
min                  0.000000                0.000000  1.620000e+02   
25%                133.000000              614.000000  5.340988e+06   
50%                371.500000              988.000000  2.551750e+07   
75%                636.000000            11000.000000  6.230944e+07   
max              23000.000000           640000.000000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count     5.043000e+03                5043.000000           5030.000000   
mean      8.366816e+04                9699.063851              1.371173   
std       1.384853e+05               18163.799124              2.013576   
min       5.000000e+00                   0.000000              0.000000   
25%       8.593500e+03                1411.000000              0.000000   
50%       3.435900e+04                3090.000000              1.000000   
75%       9.630900e+04               13756.500000              2.000000   
max       1.689764e+06              656730.000000             43.000000   

       num_user_for_reviews        budget   title_year  \
count           5022.000000  4.551000e+03  4935.000000   
mean             272.770808  3.975262e+07  2002.470517   
std              377.982886  2.061149e+08    12.474599   
min                1.000000  2.180000e+02  1916.000000   
25%               65.000000  6.000000e+06  1999.000000   
50%              156.000000  2.000000e+07  2005.000000   
75%              326.000000  4.500000e+07  2011.000000   
max             5060.000000  1.221550e+10  2016.000000   

       actor_2_facebook_likes   imdb_score  aspect_ratio  movie_facebook_likes  
count             5030.000000  5043.000000   4714.000000           5043.000000  
mean              1651.754473     6.442138      2.220403           7525.964505  
std               4042.438863     1.125116      1.385113          19320.445110  
min                  0.000000     1.600000      1.180000              0.000000  
25%                281.000000     5.800000      1.850000              0.000000  
50%                595.000000     6.600000      2.350000            166.000000  
75%                918.000000     7.200000      2.350000           3000.000000  
max             137000.000000     9.500000     16.000000         349000.000000

## Question 1
A Hollywood executive wants to know how much a R-Rated movie released after 2000 will earn. The data above is a sample of some of the movies with that ranting during that time frame, as well as other movies. How would you go about answering her question? Talk through it theoretically and then do it in code.

Formula for confidence interval: mean +- std_err*sd_multiplier, (sd_multiplier determines the percentage of the confidence interval)

for 95%, sd_multiplier is 1.96

In [10]:
#your answer here

subset = df[
    (df['content_rating']=='R') &
    (df['title_year'] >= 2000)

]

subset_mean = subset.gross.mean()
subset_std = subset.gross.std()
n = len(subset)

std_err = subset_std / (n**.5)

low_conf_int = subset_mean-std_err*1.96
high_conf_int = subset_mean+std_err*1.96

print(f"mean: {subset_mean}")
print(f"conf int: {low_conf_int}, {high_conf_int}")


mean: 27597711.116731517
conf int: 25610659.498804655, 29584762.73465838


## Question 2a

Your ability to answer the first question has the executive excited and now she has many other questions about the types of movies being made and the differences in those movies budgets and gross amounts.  Read through the questions below and determine what type of statistical test you should use for each question and write down the null and alternative hypothesis for those tests.    

- Is there a realtionship between the number of facebook likes for a cast and the box office gross of the movie?
- Do foreign films perform differently at the box office than non-foreign films?
- Of all movies created are 40% rated R?
- Is there a relationship between the language of a film and the content rating (G, PG, PG-13, R)of that film?
- Is there a relationship between the content rating of a film and its budget? 

Is there a realtionship between the number of facebook likes for a cast and the box office gross of the movie?

- Linear regression
- H_null: coeff of determiniation for facebook likes as ind, box office gross as dep is not sig different from 0
- H_alt: coeff of determiniation for facebook likes as ind, box office gross as dep is sig different from 0
- terminiology:
  - `coeff of determiniation`: r^2
  - `coeff`: value in model equation for beta coefficient, as a group called "betas" usually
  
Do foreign films perform differently at the box office than non-foreign films?
- two-sided / two-sample t-test
- H_null: there is no sig difference in the avg gross of foreign films and domestic films
- H_alt: there is a sig difference in the avg gross of foreign films and domestic films

Of all movies created are 40% rated R?
- one-sample t-test or z-score
- H_null
   - t-test: number of R-rated films is 40% of all movies
- H_alt
   - t-test: number of R_rated films is not 40% of all movies

Is there a relationship between the language of a film and the content rating (G, PG, PG-13, R)of that film?
- chi-square test
- H_null: there is no sig difference among levels of a language of a film and its content rating
- H_alt: there is a sig difference among levels of a language of a film and its content rating

Is there a relationship between the content rating of a film and its budget? 
- linear regression, ANOVA (do more than two means differ among each other sig)
- H_null: there is no sig difference among the mean of budget in groups of content ratings
- H_alt: there is a sig difference between the mean of budget in groups of content ratings

## Question 2b

Calculate the answer for the second question
- Do foreign films perform differently at the box office than non-foreign films?

In [19]:
#your answer here
from scipy.stats import ttest_ind
from scipy import stats

'''
H_null: the means of foreign and domestic films do not significantly differ
H_alt: the means of foreign and domestic films do significantly differ
alpha level= .05

one-sided or two-sided: two-sided
'''


foreign = df[df['language'] != 'English']
domestic = df[df['language'] == 'English']


ddf = (len(foreign)+len(domestic))-2

crit_below = stats.t.ppf(.025, ddf)
crit_above = stats.t.ppf(.975, ddf)

for_gross = foreign.gross
dom_gross = domestic.gross

for_gross.dropna(inplace=True)
dom_gross.dropna(inplace=True)

print(crit_below, crit_above)
ttest_ind(for_gross, dom_gross, equal_var=False)

-1.9604346909949961 1.9604346909949957


Ttest_indResult(statistic=-33.05939869299837, pvalue=1.7166302473647644e-179)

## Question 3

Now that you have answered all of those questions, the excutive wants you to create a model that predicts the money a movie will make if it is released next year in the US. She wants to use this to evaluate different scripts and then decide which one has the largest revenue potential. 

Below is a list of potential features you could use in the model. Create a new frame containing only those variables.

Would you use all of these features in the model? 

Identify which features you might drop and why.

*Remember you want to be able to use this model to predict the box office gross of a film **before** anyone has seen it.*

- `budget`: The amount of money spent to make the movie
- `actor_1_facebook_likes`: The number of likes that the most popular actor in the movie has
- `cast_total_facebook_likes`: The sum of likes for the three most popular actors in the movie
- `num_critic_for_reviews`: The number of critics who reviewed the movie
- `imdb_score`: The avg score on imdb for user reviews of the movie
- `title_year`: Year the movie was released

In [21]:
#your answer here

df_subset = df[df.content_rating.isin(['G', 'PG', 'PG-13', 'R'])]

model_data = df_subset[[
    'gross', 'budget', 'actor_1_facebook_likes', 'cast_total_facebook_likes', 
    'title_year', 'content_rating', 'imdb_score'
]]

model_data.corr()

'''
get rid of either actor_1_facebook_likes or total_cast_facebook_likes b/c of multicollineariy, which would result in coefficient values that wouldn't be stable across repeated samples 
/ result in invalid stat tests for those asvasvasv

also imdb_score or num_critic_reviews, as can't get those before movie is released
'''

gross    budget  actor_1_facebook_likes  \
gross                      1.000000  0.099710                0.146447   
budget                     0.099710  1.000000                0.019413   
actor_1_facebook_likes     0.146447  0.019413                1.000000   
cast_total_facebook_likes  0.238476  0.032363                0.946880   
title_year                 0.029171  0.043761                0.082982   
imdb_score                 0.215041  0.036205                0.102390   

                           cast_total_facebook_likes  title_year  imdb_score  
gross                                       0.238476    0.029171    0.215041  
budget                                      0.032363    0.043761    0.036205  
actor_1_facebook_likes                      0.946880    0.082982    0.102390  
cast_total_facebook_likes                   1.000000    0.109783    0.116474  
title_year                                  0.109783    1.000000   -0.159106  
imdb_score                                  0.116474   -0.159106    1.000000

## Question 4a

Create the following variables:
- `years_old`: The number of years since the film was released
- Dummy categories for each of the following ratings:
  - `G`
  - `PG`
  - `R`

Once you have those variables, create a summary output for the following OLS model:

`gross = cast_total_facebook_likes + budget + years_old + G + PG + R`

In [27]:
#your answer here

model_data['years_old'] = 2020 - model_data['title_year']

ratings = (
    pd.get_dummies(
        model_data, columns=['content_rating']
    )
    .drop(columns='content_rating_PG-13')
)

model_frame = pd.concat([model_data, ratings])

model = ols(formula="gross~cast_total_facebook_likes + budget + years_old + content_rating_G + content_rating_PG + content_rating_R", data=model_frame).fit()
model.summary()

# model_frame.sample(5)

/Users/benoren/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/Users/benoren/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:12: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  if sys.path[0] == '':


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  gross   R-squared:                       0.139
Model:                            OLS   Adj. R-squared:                  0.137
Method:                 Least Squares   F-statistic:                     100.1
Date:                Wed, 14 Oct 2020   Prob (F-statistic):          3.58e-117
Time:                        16:25:36   Log-Likelihood:                -72519.
No. Observations:                3735   AIC:                         1.451e+05
Df Residuals:                    3728   BIC:                         1.451e+05
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
=============================================================================================
                                coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------
Intercept                  5.642e+07   2.75e+06     20.522      0.000     5.1e+07    6.18e+07
cast_total_facebook_likes   823.2396     56.453     14.583      0.000     712.557     933.922
budget                        0.0258      0.005      5.449      0.000       0.017       0.035
years_old                 -5.146e+04   1.27e+05     -0.406      0.685      -3e+05    1.97e+05
content_rating_G           2.291e+07   7.18e+06      3.191      0.001    8.83e+06     3.7e+07
content_rating_PG          1.039e+07    3.3e+06      3.146      0.002    3.91e+06    1.69e+07
content_rating_R          -3.377e+07   2.41e+06    -13.986      0.000   -3.85e+07    -2.9e+07
==============================================================================
Omnibus:                     2228.542   Durbin-Watson:                   1.077
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            36066.060
Skew:                           2.534   Prob(JB):                         0.00
Kurtosis:                      17.355   Cond. No.                     1.56e+09
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.56e+09. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### Analysis

14% of the variance of `gross` is explained by the above model

the adj r^2 indicates that the number of predicted values is small enough that it's not raising the r^2 value much (since r^2 always increases with more variables in the model)

coeff analysis:
- for every unit increase in budget, gross increases by .0258
- and so on for other variables
- the variable "years old" was not stat distinguishable from 0 (at .05 level)
- other variables are stat distinguishable from 0 (at .05 level)

## Question 4b

Below is the summary output you should have gotten above.  Identify any key takeaways from it.  How ‘good’ is this model? Which features help to explain the variance in the target variable? Which do not? 


- The model we made above explains a higher percentage of the variance in gross than the same model on a different cut of the data
- There are differences in significance among the variables in the diferent models:
  - in the model below, `G` is no longer sig at .05 alpha level
  - the `years_old` variable in this model is sig at the .05 level, while above it was not

![](ols_summary.png)

In [ ]:
#your answer here

## Question 5

**Bayes Theorem**

An advertising executive is studying television viewing habits of married men and women during prime time hours. Based on the past viewing records he has determined that during prime time wives are watching television 60% of the time. It has also been determined that when the wife is watching television, 40% of the time the husband is also watching. When the wife is not watching the television, 30% of the time the husband is watching the television. Find the probability that if the husband is watching the television, the wife is also watching the television.

In [28]:
#your answer here

P_A = .6 #prob wife watching TV
P_B_given_A = .4 #prob husb watching when wife watching
P_B_given_not_A = .3 #prob of husb watching when wife is not
P_not_A = 1-P_A #prob of wife not watching

P_B = P_B_given_A*P_A +P_B_given_not_A*P_not_A

#Prob wife watching given husb watching
P_A_given_B = (P_A*P_B_given_A)/P_B
P_A_given_B

0.6666666666666666

## Question 6

Explain what a type 1 error is and how does it relate to the significance level when doing a statistical test. 

In [ ]:
'''
type 1 error is when you reject the null hypothesis when it's true 'in reality'

(boy claims there's a wolf, villiage reacts as if the null hypo of no wolf is false, when in null hypo of no wolf is in fact true)

sig level = alpha = the chance of getting a type 1 error

'''